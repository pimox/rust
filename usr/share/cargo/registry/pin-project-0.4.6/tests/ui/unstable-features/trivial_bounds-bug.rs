// NB: If you change this test, change 'trivial_bounds-feature-gate.rs' at the same time.

// trivial_bounds
// Tracking issue: https://github.com/rust-lang/rust/issues/48214
#![feature(trivial_bounds)]

use std::marker::{PhantomData, PhantomPinned};

fn phantom_pinned() {
    struct Foo(PhantomPinned);

    // bug of trivial_bounds?
    impl Unpin for Foo where PhantomPinned: Unpin {} //~ ERROR E0277

    struct Wrapper<T>(T);

    impl<T> Unpin for Wrapper<T> where T: Unpin {}

    struct Bar(PhantomPinned);

    impl Unpin for Bar where Wrapper<PhantomPinned>: Unpin {} //~ Ok

    struct WrapperWithLifetime<'a, T>(PhantomData<&'a ()>, T);

    impl<T> Unpin for WrapperWithLifetime<'_, T> where T: Unpin {}

    struct Baz(PhantomPinned);

    impl<'a> Unpin for Baz where WrapperWithLifetime<'a, PhantomPinned>: Unpin {}
    // Ok
}

fn main() {}
