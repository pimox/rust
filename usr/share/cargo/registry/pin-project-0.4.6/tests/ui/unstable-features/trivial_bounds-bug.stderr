error[E0277]: the trait bound `std::marker::PhantomPinned: std::marker::Unpin` is not satisfied
  --> $DIR/trivial_bounds-bug.rs:13:5
   |
13 |     impl Unpin for Foo where PhantomPinned: Unpin {} //~ ERROR E0277
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Unpin` is not implemented for `std::marker::PhantomPinned`
   |
   = help: the following implementations were found:
             <std::marker::PhantomPinned as std::marker::Unpin>
   = note: required by `std::marker::Unpin`
